src/aws.ts

```diff
@@ -504,6 +508,142 @@ export namespace $AWS {
       },
     });
   }
+
+  export const SDK: SDK = new Proxy<any>(
+    {},
+    {
+      get(_, serviceName: ServiceKeys) {
+        return new Proxy<any>(
+          {},
+          {
+            get: (_, methodName: string) => {
+              // if not explicitly mapped default to the lowercase service name, which is correct ~60% of the time
+              const defaultServicePrefix =
+                IAM_SERVICE_PREFIX[serviceName] ?? serviceName.toLowerCase();
+              const defaultMethod =
+                methodName.charAt(0).toUpperCase() + methodName.slice(1);
+              const defaultIamActions = [
+                `${defaultServicePrefix}:${defaultMethod}`,
+              ];
+
+              return makeIntegration<
+                `$AWS.SDK.${ServiceKeys}`,
+                (input: any) => Promise<any>
+              >({
+                kind: `$AWS.SDK.${serviceName}`,
+                native: {
+                  bind(context, args) {
+                    const options =
+                      args[1] &&
+                      (evalToConstant(args[1])?.constant as SdkCallOptions);
+
+                    if (!options) {
+                      throw new SynthError(
+                        ErrorCodes.Expected_an_object_literal,
+                        "Second argument ('options') into a SDK call is required"
+                      );
+                    } else if (typeof options !== "object") {
+                      throw new SynthError(
+                        ErrorCodes.Expected_an_object_literal,
+                        "Second argument ('options') into a SDK call must be an object"
+                      );
+                    }
+
+                    context.resource.addToRolePolicy(
+                      new aws_iam.PolicyStatement({
+                        effect: aws_iam.Effect.ALLOW,
+                        actions: options.iamActions ?? defaultIamActions,
+                        resources: options.iamResources,
+                        conditions: options.iamConditions,
```

Comment:
This is going to grant the step function the ability to invoke any resource with this policy.

Ideally we'd be able to do what CDK does and create a statement for the resource being invoked.

https://github.com/aws/aws-cdk/blob/v1.168.0/packages/@aws-cdk/aws-stepfunctions-tasks/lib/lambda/invoke.ts#L121-L126

I don't think we should create such a permissive policy by default. We either need to require at least one resource be provided (`iamResources: [string, ...string[]]`), have the developer provide their own statement outside of the call scope, or create a way to extract the arn from the call (which would require the payload be an object literal and not a reference).

To @sam-goodwin 's comment below, I wonder if we should change the IAM part to a grant method. And decouple it from the call. We could error or warn when no policy is created.

```ts
// infer the permission from the api.
AWS.SDK.CloudWatch.describeAlarms.grant({ resources: [alarm.alarmName] });
/// explicit
AWS.SDK.CloudWatch.describeAlarms.grant({ actions: ..., resources: [alarm.alarmName] });

AWS.SDK.CloudWatch.describeAlarms({ AlarmNames: [alarm.alarmName] });
```

Reply:
(cross post from discord)

Another issue I have with the approach of providing IAM permissions at request time is that the iam permissions exist outside of the scope of the call. 

All three of these blocks have the same output.

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
});
```

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
```

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, {
  ..iam stuff
});
```

and... if a user does make use of `resources` or `conditions`, any call to an api anywhere in the call stack will create a more permissive statement.

The condition below doesn't matter when myFunc is called. myFunc will grant the step function permission to call any resource with no condition.

```ts
 function myFunc() {
  $AWS.SDK.CloudWatch.describeAlarms({
     AlarmNames: [alarm.alarmName]
  });
}

StepFunction(stack, 'sfn', async () => {
      $AWS.SDK.CloudWatch.describeAlarms({
          AlarmNames: [alarm.alarmName]
      }, {
           ... some condition
      });
      myFunc();
});
```

Reply:
Does this do anything on its own? Or does it return an object that represents that policy?

```ts
AWS.SDK.CloudWatch.describeAlarms.grant({ resources: [alarm.alarmName] });
```

Reply:
Like the grant function in CDK. It adds the policy to the current step function.... Which I also don't like because it would imply positional, but would allow a higher order functions to assert permissions.

Reply:
Maybe it helps to look at how this would be used to implement the methods on actual resources?

```ts
class Queue<T> {
  private readonly sqs;
  constructor(scope: Construct, id: string) {
    this.queue = new aws_sqs.Queue(this, "Queue");
    this.sqs = $AWS.SQS.grant(this.queue.grantRead)
  }

  public sendMessage(message: T): Promise<void> {
    await this.sqs.sendMessage({
      QueueArn: this.resource.arn,
      MessageBody: JSON.stringify(message)
    });
  }
}
```

Reply:
ok, so this makes a scoped "resource" (`this.sqs`) that can grant any caller of `Queue.sendMessage`, the permission to call  sendMessage on `this.queue`.

What would this look like to give access to an array of queues? can it work with strings and not constructs? and do we need to understand all of the `grant*` methods in CDK?

Reply:
What is the least we could do to ship this? Could we ship without any inferred IAM and only with either an explicit option or   using the CDK options. Both options with documentation.

Reply:
So to add my 2cents here:

What I found very helpful from the cdk stepfunctions modules is that they add iam permissions automatically. Across all the state machines we have, there were less than a handful where we had to add explicit permissions ourselves. I kind of wanted to go into a similar direction here as well. (Adding the IAM prefix map makes it even more seamless than the cdk stepfunctions module). The use of the `iamResources` came directly from the [`CallAwsService` task in the stepfunctions cdk](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions_tasks.CallAwsService.html#iamresources).

I do however also find the current approach unsatisfactory. But I guess what I would expect from a solution is the following:

1. infer as much as is reliably doable:
  That basically means the iam action as the resources are already infeasible to do (there are thousands) and most services don't use arns in their sdk methods but rather names so constructing the arn from the call is another difficult tasks.
2. no magic
  replacing the `AlarmName` with a `aws_cloudwatch.Alarm` class in the sdk call input and then inferring the required `grant...()` method from the sdk call is something that IMO should be reserved for the higher level integrations. As mentioned yesterday I think we should of this as a sort of L1 construct. To avoid duplicate IAM statements we could build a cache of added statements indexed by something like the object's hash. But I think that is an optimization that can be added later. Whether or not conditions come into effect or not is probably nothing we should solve as this goes where deep into IAM land and could lead to nasty edge cases.

3. make the requirement for iam explicit and allow full control if wanted
  Making the requirement for the iam resources explicit via a required type enforces users to actually specify them instead of experiencing opaque 403s during testing. An escape hatch for the inferred iam action is definitely needed and adding conditions allows even more control (that is something that the stepfunctions cdk is missing IMO)

With these things in mind I think what could make sense it one of the following:

## Adding a second parameter

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  AlarmNames: [alarm.alarmName]
}, new iam.PolicyStatement({resources: [alarm.alarmArn]})); // or use the existing object literal but I like this better
```

## adding a chained call

```ts
$AWS.SDK.CloudWatch.withIAM({
  resources: [alarm.alarmArn]
}).describeAlarms({
  AlarmNames: [alarm.alarmName]
});

$AWS.SDK.CloudWatch.withPolicy(
  new iam.PolicyStatement({resources: [alarm.alarmArn]})
).describeAlarms({
  AlarmNames: [alarm.alarmName]
});

$AWS.SDK.CloudWatch.grant({
  resources: [alarm.alarmArn]
}).describeAlarms({
  AlarmNames: [alarm.alarmName]
});
```

## nesting the sdk parameters and other things all into the first argument (Similar to what the cdk does)

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  params: {
    AlarmNames: [alarm.alarmName]
  },
  iamResources: [alarm.alarmArn],
});

// or use a PolicyStatement instead
$AWS.SDK.CloudWatch.describeAlarms({
  params: {
    AlarmNames: [alarm.alarmName]
  },
  policy:  new iam.PolicyStatement({resources: [alarm.alarmArn]}),
});
```

Reply:
I think I like the last option the most

```ts
$AWS.SDK.CloudWatch.describeAlarms({
  params: {
    AlarmNames: [alarm.alarmName]
  },
  policy:  new iam.PolicyStatement({resources: [alarm.alarmArn]}),
});
```

Reply:
I am OK with following CDK for precedence, thanks for pointing that out and for such a detailed response.

I prefer option 1 with three changes:
1. don't use the policy statement, agree that re-use isn't helpful when the contract is different
2. Validate that a resource is provided or an explicit "*"
3. create an "options" object which contains more than just IAM in it for future configuration (ex: retry for sfn)
4. Require the iam object always be provided, even if that means duplication, we can relax this later or higher order functions will solve for this.

Because:
1. It separates the runtime and synth time configuration
2. It doesn't require a constant object literal be used for the runtime payload in order to extract the synth time values
3. users cannot shoot themselves in the foot by providing an empty resources array implicitly
5. the chained interface is a little cluncky too and doesn't add much value when the values are required.
6. We did something similar to option 3 for Sfn's invocation method and I don't like having to put the payload in an named object parameter every time.